include/deal.II/lac/solver_cg.h

```diff
@@ -389,6 +468,725 @@ SolverCG<VectorType>::compute_eigs_and_cond(
+    {
+      using Number = typename VectorType::value_type;
+
+      IterationWorker(const MatrixType &        A,
+                      const PreconditionerType &preconditioner,
+                      const bool                flexible,
+                      VectorMemory<VectorType> &memory,
+                      VectorType &              x)
+        : IterationWorkerBase<VectorType, MatrixType, PreconditionerType>(
+            A,
+            preconditioner,
+            flexible,
+            memory,
+            x)
+      {}
+
+      // This is the main iteration function, that will use some of the
+      // specialized functions below
+      void
+      do_iteration(const unsigned int iteration_index)
+      {
+        std::array<VectorizedArray<Number>, 7> vectorized_sums = {};
+
+        this->A.vmult(
+          this->v,
+          this->p,
+          [&](const unsigned int begin, const unsigned int end) {
+            operation_before_loop(iteration_index, begin, end);
+          },
+          [&](const unsigned int begin, const unsigned int end) {
+            operation_after_loop(begin, end, vectorized_sums);
+          });
+
+        std::array<Number, 7> scalar_sums;
+        for (unsigned int i = 0; i < 7; ++i)
+          scalar_sums[i] = vectorized_sums[i][0];
+        for (unsigned int l = 1; l < VectorizedArray<Number>::size(); ++l)
+          for (unsigned int i = 0; i < 7; ++i)
+            scalar_sums[i] += vectorized_sums[i][l];
+
+        Utilities::MPI::sum(dealii::ArrayView<const Number>(scalar_sums.data(),
+                                                            7),
+                            this->r.get_mpi_communicator(),
+                            dealii::ArrayView<Number>(scalar_sums.data(), 7));
+
+        this->previous_alpha = this->alpha;
+        this->previous_beta  = this->beta;
+
+        const Number p_dot_A_dot_p = scalar_sums[0];
+        Assert(std::abs(p_dot_A_dot_p) != 0., ExcDivideByZero());
+
+        const Number previous_r_dot_preconditioner_dot_r = scalar_sums[6];
+        this->alpha = previous_r_dot_preconditioner_dot_r / p_dot_A_dot_p;
+        this->residual_norm = std::sqrt(
+          scalar_sums[3] +
+          this->alpha * (-2. * scalar_sums[2] + this->alpha * scalar_sums[1]));
+
+        this->r_dot_preconditioner_dot_r =
+          previous_r_dot_preconditioner_dot_r +
+          this->alpha * (-2. * scalar_sums[4] + this->alpha * scalar_sums[5]);
+
+        this->beta = this->r_dot_preconditioner_dot_r /
+                     previous_r_dot_preconditioner_dot_r;
+      }
+
+      // Function that we use if the PreconditionerType implements an apply()
+      // function
+      template <typename U = void>
+      typename std::enable_if<has_apply<PreconditionerType>, U>::type
+      operation_before_loop(const unsigned int iteration_index,
+                            const unsigned int start_range,
+                            const unsigned int end_range) const
```

Comment:
This here is one example (out of 3 such functions) where I apply this detection.